### ISSUE TITLE
Obtain agent ID from space

### ISSUE THREAD

**Issue by xaviarias** [sarl/sarl#291]
Hi,

In order to identify which agent is invoking a space method from its implementation, is that possible in SARL to obtain the caller ID?

Thanks

**Comment by gallandarakhneorg** [comment]
Hi Xavi,

I think I'm needing more details about that you want to do. But, I may give you the following guidelines.

First, the space implementation should have the identity of the caller in the function prototype:
```[java]
class MySpaceImpl implements MySpace {
  public void doSomething(UUID agentID, Object... params) {
  }
}
```

Then, the best practice seems to define a capacity and a skill for accessing the space (in which the agent identity are removed from the parameters):
```
capacity MyCapacity {
    def doSomething(params : Object*)
}
skill MySkill implements MyCapacity {
   val space : MySpace
   new (space : MySpace) {
      this.space = space
   }
   def doSomething(params : Object*) {
      space.doSomething(owner.ID, params)
   }
}
```

Then in your agent, you could call the capacity as:
```
agent MyAgent {
   uses DefaultContextInteraction, MyCapacity
   on Initialize {
      var space = defaultContext.getSpace(occurrence.parameters.get(0))
      setSkill( new MySkill(space))
   }
   on Something {
      doSometging(param1, param2, param3)
   }
}
```

It is corresponding to your expected behavior?

**Comment by xaviarias** [comment]
Hi Stéphane,

Thanks for the details, there's no problem regarding the skills, capacities, etc. I'm able to run the agent platform but was a question of design. I'm trying to abstract the agents from a e-voting platform which I have implemented as a space following your guidelines from a previous issue (#225) I opened some time ago.

The concept is similar to the Agents & Artifacts metamodel implemented by the Cartago project (http://cartago.sourceforge.net/). In this approach, artifacts have operations, properties an can trigger events. In addition, artifact implementations can identify their operation callers in order to prevent agents to manipulate the artifact using someone else's identity.

The idea using a space if very similar: agents "register" in the space in order to receive its events, but cannot emit any (I have implemented this using a private field OpenEventSpace inside the space class), and can invoke space operations like an artifact. The question was regarding the identity parameter, since Cartago allows to determine the caller implicitly.

Best regards
